bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BValueArray.java

```diff
@@ -0,0 +1,573 @@
+import org.ballerinalang.persistence.serializable.reftypes.impl.SerializableBRefArray;
+import org.ballerinalang.util.BLangConstants;
+import org.ballerinalang.util.exceptions.BallerinaException;
+import org.wso2.ballerinalang.compiler.util.BArrayState;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import static org.ballerinalang.model.util.FreezeUtils.handleInvalidUpdate;
+import static org.ballerinalang.model.util.FreezeUtils.isOpenForFreeze;
+
+/**
+ * @since 0.87
+ */
+public class BValueArray extends BNewArray implements Serializable {
+
+    BRefType<?>[] refValues;
+    private long[] intValues;
+    private int[] booleanValues;
+    private byte[] byteValues;
+    private double[] floatValues;
+    private String[] stringValues;
+
+    public BType elementType;
+
+    //------------------------ Constructors -------------------------------------------------------------------
+
+    public BValueArray(BRefType<?>[] values, BType type) {
+        this.refValues = values;
+        super.arrayType = type;
+        this.size = values.length;
+        if (type != null) {
+            if (type.getTag() == TypeTags.ARRAY_TAG) {
+                this.elementType = ((BArrayType) type).getElementType();
+            } else {
+                this.elementType = type;
+            }
+        }
+    }
+
+    public BValueArray(long[] values) {
+        this.intValues = values;
+        this.size = values.length;
+        super.arrayType = new BArrayType(BTypes.typeInt);
+        this.elementType = BTypes.typeInt;
+    }
+
+    public BValueArray(int[] values) {
+        this.booleanValues = values;
+        this.size = values.length;
+        super.arrayType = new BArrayType(BTypes.typeBoolean);
+        this.elementType = BTypes.typeBoolean;
+    }
+
+    public BValueArray(byte[] values) {
+        this.byteValues = values;
+        this.size = values.length;
+        super.arrayType = new BArrayType(BTypes.typeByte, size);
+        this.elementType = BTypes.typeByte;
+    }
+
+    public BValueArray(double[] values) {
+        this.floatValues = values;
+        this.size = values.length;
+        super.arrayType = new BArrayType(BTypes.typeFloat);
+        this.elementType = BTypes.typeFloat;
+    }
+
+    public BValueArray(String[] values) {
+        this.stringValues = values;
+        this.size = values.length;
+        super.arrayType = new BArrayType(BTypes.typeString);
+        this.elementType = BTypes.typeString;
+    }
+
+    public BValueArray(BType type) {
+        if (type.getTag() == TypeTags.INT_TAG) {
+            intValues = (long[]) newArrayInstance(Long.TYPE);
+            super.arrayType = new BArrayType(BTypes.typeInt);
```

Comment:
we can call the other constructor here by passing `(long[]) newArrayInstance(Long.TYPE);`. That will reduce the code duplication.